File: plugins/gui/frontend/src/logic/mergeData.js
const mergedData = (data, schema) => {

  const isSubsetValue = (subsetValue, supersetValue) => {
    if (typeof (subsetValue) !== typeof (supersetValue)) return false;
    if (Array.isArray(subsetValue)) {
      if (subsetValue.length > supersetValue.length) return false;
      for (let i = 0; i < subsetValue.length; i += 1) {
        let found = false;
        supersetValue.forEach((supersetItem) => {
          if (isSubsetValue(subsetValue[i], supersetItem)) found = true;
        });
        if (!found) return false;
      }
      return true;
    }
    if (typeof (subsetValue) === 'object') {
      // eslint-disable-next-line no-use-before-define
      return isSubset(subsetValue, supersetValue);
    }
    return (subsetValue === supersetValue);
  };

  const isSubset = (subset, superset) => {
    let result = true;
    Object.entries(subset).forEach(([key, value]) => {
      if (key !== 'adapters') {
        if (!result || (value && !superset[key]) || !isSubsetValue(value, superset[key])) {
          result = false;
        }
      }
    });
    return result;
  };
  // returns a "merged" version of the data:
  // "if a row is a full subset of another row, do not show it"
  const localData = [...data];
  const totalExcludedSet = new Set();
  const stringFields = schema.items.filter((x) => x.type === 'string' && !x.name.includes('.'));

  // out of all fields, on the topmost level, which are scalars (specifically a string)
  // those that are unique (and non-null) are never a superset or a subset of any other row
  // this code will find those unique ones and add them to 'totalExcludedSet', so they will
  // not take place in the rest of the algorithm.
  // this is O(n)
  stringFields.forEach((schemaItem) => {
    const fieldName = schemaItem.name;

    const excludedSet = {};
    const valuesSet = new Set();

    localData.map((d) => d[fieldName]).forEach((dataItem, index) => {
      if (!dataItem) return;

      if (valuesSet.has(dataItem)) {
        delete excludedSet[dataItem];
      } else {
        excludedSet[dataItem] = index;
        valuesSet.add(dataItem);
      }
    });

    Object.values(excludedSet).forEach((x) => totalExcludedSet.add(x));
  });

  // this is the lousy part of this algorithm, it is in O(n^2)
  // where (n) = amount of row - amount of rows eliminated in previous steps
  // Goes over any unique pair and check if one is a subset of the other, and if it is,
  // hide show the subset.
  const result = Array.from(totalExcludedSet).map((index) => localData[index]);
  localData.forEach((subset, index) => {
    if (subset === undefined || totalExcludedSet.has(index)) return;
    let found = false;
    for (let i = 0; i < result.length; i += 1) {
      const superset = result[i];
      if (superset === undefined) return;
      let mergeAdapters = false;

      if (isSubset(superset, subset)) {
        result[i] = subset;
        found = true;
        mergeAdapters = true;
      } else if (isSubset(subset, superset)) {
        found = true;
        mergeAdapters = true;
      }

      if (mergeAdapters) {
        result[i].adapters = [...new Set([
          ...superset.adapters, ...subset.adapters, ...result[i].adapters,
        ])];
      }
    }

    if (!found) {
      result.push(subset);
    }
  });

  return result;
};


export default mergedData;

File: axonius-libs/src/libs/axonius-py/axonius/utils/merge_data.py
from axonius.utils.gui_helpers import is_subjson, parse_entity_fields


# pylint:disable=too-many-statements, too-many-locals, too-many-branches, too-many-nested-blocks
def merge_entities_fields(entities_data: dict, fields: dict):
    """
        find all entities that are subset of other entities, and merge them.

        this code is a port of the mergeData.js file.

    :param entities_data:  the entities to do the merge on
    :param fields:  schema dictionary mapped by name
    :return: entities merged by subsets
    """

    string_fields = [item['name'] for item in fields.values() if item['type'] == 'string' and '.' not in item['name']]
    local_data = [parse_entity_fields(entity_data, fields.keys()) for entity_data in entities_data]

    total_excluded_set = set()

    for field_name in string_fields:
        excluded_set = {}
        values_set = set()

        normalizes_data = [item.get(field_name) for item in local_data if item]

        for index, data_item in enumerate(normalizes_data):
            if not data_item:
                continue

            if data_item in values_set:
                excluded_set.pop(data_item, None)
            else:
                excluded_set[data_item] = index
                values_set.add(data_item)

        total_excluded_set = set(total_excluded_set).union(excluded_set.values())

    result = [local_data[index] for index in total_excluded_set]
    for index, subset in enumerate(local_data):

        if subset is None or index in total_excluded_set:
            continue
        found = False

        for superset_index, superset in enumerate(result):
            if superset is None:
                continue

            merge_adapters = False

            if is_subjson(superset, subset):
                result[superset_index] = subset
                found = True
                merge_adapters = True
            elif is_subjson(subset, superset):
                found = True
                merge_adapters = True

            if merge_adapters and\
                    'adapters' in superset and\
                    'adapters' in subset and\
                    'adapters' in result[superset_index]:
                result[superset_index]['adapters'] = sorted(list(
                    set(superset['adapters']).union(subset['adapters']).union(result[superset_index]['adapters'])
                ))

        if not found:
            result.append(subset)

    if result and 'adapters' in result[0]:
        result.sort(key=lambda x: x['adapters'], reverse=False)

    return result
